Document precompiled Groovy script plugins (#12704)
gradle/gradle: subprojects/docs/src/docs/userguide/extending-gradle/custom_plugins.adoc

include::sample[dir="snippets/customPlugin/kotlin/consumer",files="settings.gradle.kts[tags=use-legacy-plugin]"]
====

[[sec:precompiled_plugins]]
== Precompiled script plugins

> In addition to plugins written as regular projects, Gradle allows you to provide the plugin logic
> in either Groovy or Kotlin DSL as precompiled script plugins. You write these as `\*.gradle` files in `src/main/groovy`
> directory or `*.gradle.kts` files in `src/main/kotlin` directory.

Reviewer: ```suggestion
In addition to plugins written as standalone projects, Gradle also allows you to provide build logic written in either Groovy or Kotlin DSLs as precompiled script plugins. You write these as `\*.gradle` files in `src/main/groovy`
directory or `*.gradle.kts` files in `src/main/kotlin` directory.
```

> Precompiled script plugins are compiled as part of a regular source set and placed on the build classpath or packaged in
> a binary plugin, depending on what type of project they are in. For all intents and purposes, they are binary plugins,
> particularly as they can be applied by plugin ID, just like a regular binary plugin. In fact, the plugin metadata for
> them is generated via implicitly applied <<java_gradle_plugin#java_gradle_plugin,Gradle Plugin Development Plugin>>.

Reviewer: ```suggestion
Precompiled script plugins are compiled into class files and packaged into a jar. For all intents and purposes, they are binary plugins and can be applied by plugin ID, tested and published as binary plugins. In fact, the plugin metadata for
them is generated using the <<java_gradle_plugin#java_gradle_plugin,Gradle Plugin Development Plugin>>.
```

[NOTE]
====
Kotlin DSL precompiled script plugins built with Gradle 6.0 cannot be used with earlier versions of Gradle.
This limitation will be lifted in a future version of Gradle.

> Groovy DSL precompiled script plugins are available starting with Gradle 6.4 and they can not be used with earlier versions.

Reviewer: I think it's OK to say this is the limitation for Groovy DSL until we're sure what the backwards guarantees will be in the future.
